examples/DataApi-NFT/template/modules/shared_data/src/main/scala/com/my/nft_example/shared_data/Validations.scala

```diff
@@ -0,0 +1,94 @@
+package com.my.nft_example.shared_data
+
+import cats.data.NonEmptySet
+import cats.effect.IO
+import cats.implicits.{catsSyntaxApply, toFoldableOps, toTraverseOps}
+import com.my.nft_example.shared_data.Data.{MintCollection, MintNFT, State, TransferCollection, TransferNFT}
+import com.my.nft_example.shared_data.TypeValidators.{validateIfCollectionIsUnique, validateIfFromAddressIsTheCollectionOwner, validateIfFromAddressIsTheNFTOwner, validateIfNFTIdIsUnique, validateIfNFTUriIsUnique, validateIfNFTUriIsValid, validateIfProvidedCollectionExists, validateIfProvidedNFTExists, validateProvidedAddress}
+import org.tessellation.currency.dataApplication.dataApplication.DataApplicationValidationErrorOr
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.signature.SignatureProof
+
+object Validations {
+  def mintCollectionValidations(update: MintCollection, state: State): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateCollection = validateIfCollectionIsUnique(update, state)
+    IO {
+      validateCollection
+    }
+  }
+
+  def mintNFTValidations(update: MintNFT, state: State): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateNFTUrlValid = validateIfNFTUriIsValid(update)
+    val validateUniqueNFTURI = validateIfNFTUriIsUnique(update, state)
+    val validateUniqueNFTId = validateIfNFTIdIsUnique(update, state)
+
+    IO {
+      validateNFTUrlValid.productR(validateUniqueNFTURI).productR(validateUniqueNFTId)
+    }
+  }
+
+  def transferCollectionValidations(update: TransferCollection, state: State): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateProvidedCollection = validateIfProvidedCollectionExists(update, state)
+    val validateFromAddress = validateIfFromAddressIsTheCollectionOwner(update, state)
+
+    IO {
+      validateProvidedCollection.productR(validateFromAddress)
+    }
+
+  }
+
+  def transferNFTValidations(update: TransferNFT, state: State): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateProvidedNFT = validateIfProvidedNFTExists(update, state)
+    val validateFromAddress = validateIfFromAddressIsTheNFTOwner(update, state)
+
+    IO {
+      validateProvidedNFT.productR(validateFromAddress)
+    }
+  }
+
+  def mintNFTValidationsWithSignature(update: MintNFT, proofs: NonEmptySet[SignatureProof], state: State)(implicit sp: SecurityProvider[IO]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateAddress = proofs
+      .map(_.id)
+      .toList
+      .traverse(_.toAddress[IO])
+      .map(validateProvidedAddress(_, update.owner))
+
+    val validations = mintNFTValidations(update, state)
```

Comment:
Again similar comment as I made before, some of the IO constructions inside the validator functions (where they don't use IO, are unnecessary and can be removed. 

To give an example here as it relates to option, this is a bit like having one function that already returns Option[String] (out of your control) -- and then you implement another function that returns Some("astring") which is NEVER a None type.

so in pseudocode
option1 = fnReturningOptionWhichMightBeNone()
option2 = Some("hey")

Then in this function here, you're doing option1.flatMap{o => option2.map{v => o +v}}

When you could have just done

option1.map{o => o + v}